java/java-psi-api/src/com/intellij/psi/util/OrderedBinaryExpression.java

```diff
@@ -0,0 +1,122 @@
+    PsiBinaryExpression expression = as(node, PsiBinaryExpression.class);
+    if (expression == null || firstClass == null) return null;
+    O leftOperand = as(expression.getLOperand(), firstClass);
+    O rightOperand = as(expression.getROperand(), firstClass);
+    if (leftOperand != null && rightOperand == null) {
+      return new OrderedBinaryExpression<>(leftOperand, expression.getOperationTokenType(), PsiUtil.skipParenthesizedExprDown(expression.getROperand()));
+    }
+    if (leftOperand == null && rightOperand != null) {
+      IElementType mirroredOperator = mirrorOperator(expression);
+      return new OrderedBinaryExpression<>(rightOperand, mirroredOperator, PsiUtil.skipParenthesizedExprDown(expression.getLOperand()));
+    }
+    return null;
+  }
+
+  /**
+   * Return the items of a binary expression in the order it is specified. It reverses the operator if needed.
+   *
+   * @param <F>          the required expression type
+   * @param <S>          the required expression type
+   * @param node         the supposed binary expression
+   * @param firstClass   the class representing the required expression type
+   * @param secondClass  the class representing the required expression type
+   * @return the items of a binary expression in the order it is specified. It reverses the operator if needed.
+   */
+  public static <F extends PsiExpression, S extends PsiExpression> OrderedBinaryExpression<F, S> from(final PsiExpression node, final Class<F> firstClass, final Class<S> secondClass) {
+    PsiBinaryExpression expression = as(node, PsiBinaryExpression.class);
+    if (expression == null || firstClass == null || secondClass == null) return null;
+    F leftFirst = as(expression.getLOperand(), firstClass);
+    S rightSecond = as(expression.getROperand(), secondClass);
+    if (leftFirst != null && rightSecond != null) {
+      return new OrderedBinaryExpression<>(leftFirst, expression.getOperationTokenType(), rightSecond);
+    }
+    IElementType mirroredOperator = mirrorOperator(expression);
+    if (mirroredOperator != null) {
+      F rightFirst = as(expression.getROperand(), firstClass);
+      S leftSecond = as(expression.getLOperand(), secondClass);
+
+      if (rightFirst != null && leftSecond != null) {
+        return new OrderedBinaryExpression<>(rightFirst, mirroredOperator, leftSecond);
+      }
+    }
+    return null;
+  }
+
+  private static <T extends PsiExpression> T as(PsiExpression input, Class<T> aClass) {
+    return tryCast(PsiUtil.skipParenthesizedExprDown(input), aClass);
+  }
+
+  private static IElementType mirrorOperator(final PsiBinaryExpression expression) {
+    if (Arrays.asList(
+      JavaTokenType.AND,
+      JavaTokenType.ANDAND,
+      JavaTokenType.OROR,
+      JavaTokenType.EQEQ,
+      JavaTokenType.NE,
+      JavaTokenType.OR,
+      JavaTokenType.PLUS,
+      JavaTokenType.ASTERISK,
+      JavaTokenType.XOR).contains(expression.getOperationTokenType())) {
```

Comment:
Let's extract `expression.getOperationTokenType()` to a variable. It's not like a simple getter.

Reply:
Or better pass it as a parameter to this method instead of `expression`.

Reply:
Done